llvm/lib/MC/MCCASObjectV1.cpp

```diff
@@ -753,36 +753,302 @@ DebugAbbrevSectionRef::materialize(MCCASReader &Reader,
   return Size;
 }
 
+struct LineTablePrologue {
+  uint64_t Length;
+  uint16_t Version;
+  uint32_t PrologueLength;
+  uint64_t Offset;
+  uint8_t OpcodeBase;
+  dwarf::DwarfFormat Format;
+};
+
+static Error
+getLineTableLengthInfoAndVersion(DWARFDataExtractor &LineTableDataReader,
+                                 uint64_t *OffsetPtr,
+                                 LineTablePrologue &Prologue) {
+  Error Err = Error::success();
+  // From DWARF 5 section 7.4:
+  // In the 32-bit DWARF format, an initial length field [...] is an unsigned
+  // 4-byte integer (which must be less than 0xfffffff0);
+  auto Length =
+      LineTableDataReader.getRelocatedValue(4, OffsetPtr, nullptr, &Err);
+  if (Err)
+    return Err;
+  if (Length >= 0xfffffff0)
+    return createStringError(inconvertibleErrorCode(),
+                             "DWARF input is not in the 32-bit format");
+  Prologue.Length = Length;
+  Prologue.Format = llvm::dwarf::DWARF32;
+  auto Version = LineTableDataReader.getU16(OffsetPtr, &Err);
+  if (Err)
+    return Err;
+  if (Version >= 5)
+    return createStringError(inconvertibleErrorCode(),
+                             "DWARF 5 and above is not currently supported");
+  Prologue.Version = Version;
+  // Since we do not support 64 bit DWARF, the prologue length is 4 bytes in
+  // size.
+  auto PrologueLength =
+      LineTableDataReader.getRelocatedValue(4, OffsetPtr, nullptr, &Err);
+  if (Err)
+    return Err;
+
+  Prologue.PrologueLength = PrologueLength;
+  Prologue.Offset = *OffsetPtr + PrologueLength;
+  return Err;
+}
+
+static Expected<LineTablePrologue>
+parseLineTableHeaderAndSkip(DWARFDataExtractor &LineTableDataReader) {
+  uint64_t Offset = 0;
+  uint64_t *OffsetPtr = &Offset;
+  LineTablePrologue Prologue;
+  Error Err = getLineTableLengthInfoAndVersion(LineTableDataReader, OffsetPtr,
+                                               Prologue);
+  if (Err)
+    return Err;
+  // Parse Minimum instruction length.
+  LineTableDataReader.getRelocatedValue(1, OffsetPtr, nullptr, &Err);
+  // Parse Maximum Operands Per Instruction, if it exists.
+  if (Prologue.Version >= 4)
+    LineTableDataReader.getRelocatedValue(1, OffsetPtr, nullptr, &Err);
+  // Parse DefaultIsStmt, LineBase, and LineRange.
+  LineTableDataReader.getRelocatedValue(1, OffsetPtr, nullptr, &Err);
+  LineTableDataReader.getRelocatedValue(1, OffsetPtr, nullptr, &Err);
+  LineTableDataReader.getRelocatedValue(1, OffsetPtr, nullptr, &Err);
+  // Parse OpcodeBase.
+  Prologue.OpcodeBase =
+      LineTableDataReader.getRelocatedValue(1, OffsetPtr, nullptr, &Err);
```

Comment:
I should use a getU8 here, but a BinaryStreamReader cannot extract ULEB values, so I am still gonna stick with the DWARFDataExtractor. Also, this is what is used by the Line Table Parser in DWARFDebugLine.cpp

Reply:
>Also, this is what is used by the Line Table Parser in DWARFDebugLine.cpp

I suspect they _need_ to know about relocations though.

> t a BinaryStreamReader cannot extract ULEB values

LLVM has free functions to handle those with streams: https://llvm.org/doxygen/LEB128_8h_source.html

Reply:
I don't really see a difference, why does it matter if we use a DWARFDataExtractor which can handle ULEBs and using a BinaryStreamReader and include the LEB128 header? I can replace the getRelocatedValue(1, ...) with a getU8 but I think it doesn't matter either way

Reply:
Its APIs seem a lot more complex (with the whole Cursor business), but I'll leave it up to you.

Regarding the "Relocated" functions, I just wanted to make sure we're on the same page: if we use those functions, we'd be storing the relocated values in the CAS, instead of the pre-relocation values? If this is true, isn't it a problem?

Reply:
> Its APIs seem a lot more complex (with the whole Cursor business), but I'll leave it up to you.

In my mind I find it easier to work with that BinaryStreamReader tbh.

> Regarding the "Relocated" functions, I just wanted to make sure we're on the same page: if we use those functions, we'd be storing the relocated values in the CAS, instead of the pre-relocation values? If this is true, isn't it a problem?

Yeah I agree, I just used that function because it takes a size of bytes to read, I later realized there was a getU<size> variant as well. The relocations are zero-ed out, it is a non-issue